### ISSUE TITLE
Installing and running with CUDA in Windows 10 raises Pickle error

### ISSUE THREAD

**Issue by layadas** [liznerski/fcdd#50]
I am working on a computer with Windows 10 with one GPU. I have the following situation:

- I install using the command `pip install . --user` from the `python` directory.
- Then I use the command `python runners/run_mvtec.py` from the `fcdd` directory that appears after installation. **_This causes an error with CUDA_**:
```
Traceback (most recent call last):
  File "C:\Work\IMAGE\New folder\ADProject1\fcdd-master\python\fcdd\runners\run_mvtec.py", line 19, in <module>
    runner.run()
  File "C:\Users\laydas\AppData\Roaming\Python\Python39\site-packages\fcdd\runners\bases.py", line 203, in run
    self.run_classes(**vars(self.args))
  File "C:\Users\laydas\AppData\Roaming\Python\Python39\site-packages\fcdd\runners\bases.py", line 221, in run_classes
    res = self.run_seeds(
  File "C:\Users\laydas\AppData\Roaming\Python\Python39\site-packages\fcdd\runners\bases.py", line 181, in run_seeds
    res = self.run_one(
  File "C:\Users\laydas\AppData\Roaming\Python\Python39\site-packages\fcdd\runners\bases.py", line 100, in run_one
    setup = trainer_setup(
  File "C:\Users\laydas\AppData\Roaming\Python\Python39\site-packages\fcdd\training\setup.py", line 113, in trainer_setup
    net = net.to(device)
  File "C:\Users\laydas\AppData\Roaming\Python\Python39\site-packages\torch\nn\modules\module.py", line 852, in to
    return self._apply(convert)
  File "C:\Users\laydas\AppData\Roaming\Python\Python39\site-packages\torch\nn\modules\module.py", line 530, in _apply
    module._apply(fn)
  File "C:\Users\laydas\AppData\Roaming\Python\Python39\site-packages\torch\nn\modules\module.py", line 530, in _apply
    module._apply(fn)
  File "C:\Users\laydas\AppData\Roaming\Python\Python39\site-packages\torch\nn\modules\module.py", line 552, in _apply
    param_applied = fn(param)
  File "C:\Users\laydas\AppData\Roaming\Python\Python39\site-packages\torch\nn\modules\module.py", line 850, in convert
    return t.to(device, dtype if t.is_floating_point() or t.is_complex() else None, non_blocking)
  File "C:\Users\laydas\AppData\Roaming\Python\Python39\site-packages\torch\cuda\__init__.py", line 166, in _lazy_init
    raise AssertionError("Torch not compiled with CUDA enabled")
AssertionError: Torch not compiled with CUDA enabled
```

- To correct this situation, I install the latest PyTorch.
- I then execute `python runners/run_mvtec.py` from the `fcdd` folder. Now **_this causes an error associated with a `pickle` command of a `Lambda` function_**.

Here is the complete output after executing `python runners/run_mvtec.py`:
```
Files already downloaded.
Loading dataset from C:\Work\IMAGE\New folder\ADProject\fcdd_master\python\data\datasets\mvtec\admvtec_240x240.pt...
Dataset complete.
Files already downloaded.
Loading dataset from C:\Work\IMAGE\New folder\ADProject\fcdd_master\python\data\datasets\mvtec\admvtec_240x240.pt...
Dataset complete.
Downloading: "https://download.pytorch.org/models/vgg11_bn-6002323d.pth" to C:\ProgramData\Anaconda3\envs\ADProject\lib\site-packages\fcdd\models\..\..\..\data\models\vgg11_bn-6002323d.pth
100%|██████████| 507M/507M [00:17<00:00, 30.0MB/s]
##### NET RECEPTION {'n': 28, 'j': 8, 'r': 62, 's': 3.5, 'img_shape': (3, 224, 224)} #####
Successfully saved code at C:\Work\IMAGE\New folder\ADProject\fcdd_master\python\data\results\fcdd_20221025165156_mvtec_\normal_0\it_0\.\src.tar.gz
Generating dataset preview...
Plotting ROC for completed classes up to 0...
Traceback (most recent call last):
  File "C:\ProgramData\Anaconda3\envs\ADProject\lib\code.py", line 90, in runcode
    exec(code, self.locals)
  File "<input>", line 1, in <module>
  File "C:\Program Files\JetBrains\PyCharm 2021.3.2\plugins\python\helpers\pydev\_pydev_bundle\pydev_umd.py", line 198, in runfile
    pydev_imports.execfile(filename, global_vars, local_vars)  # execute the script
  File "C:\Program Files\JetBrains\PyCharm 2021.3.2\plugins\python\helpers\pydev\_pydev_imps\_pydev_execfile.py", line 18, in execfile
    exec(compile(contents+"\n", file, 'exec'), glob, loc)
  File "C:/Work/IMAGE/New folder/ADProject/fcdd_master/python/fcdd/runners/run_mvtec.py", line 19, in <module>
    runner.run()
  File "C:\ProgramData\Anaconda3\envs\ADProject\lib\site-packages\fcdd\runners\bases.py", line 203, in run
    self.run_classes(**vars(self.args))
  File "C:\ProgramData\Anaconda3\envs\ADProject\lib\site-packages\fcdd\runners\bases.py", line 221, in run_classes
    res = self.run_seeds(
  File "C:\ProgramData\Anaconda3\envs\ADProject\lib\site-packages\fcdd\runners\bases.py", line 181, in run_seeds
    res = self.run_one(
  File "C:\ProgramData\Anaconda3\envs\ADProject\lib\site-packages\fcdd\runners\bases.py", line 100, in run_one
    setup = trainer_setup(
  File "C:\ProgramData\Anaconda3\envs\ADProject\lib\site-packages\fcdd\training\setup.py", line 122, in trainer_setup
    images = ds.preview(20, classes=[0, 1] if supervise_mode != "unsupervised" else [0], train=True)
  File "C:\ProgramData\Anaconda3\envs\ADProject\lib\site-packages\fcdd\datasets\bases.py", line 99, in preview
    for xb, yb, gtsb in loader:
  File "C:\ProgramData\Anaconda3\envs\ADProject\lib\site-packages\torch\utils\data\dataloader.py", line 444, in __iter__
    return self._get_iterator()
  File "C:\ProgramData\Anaconda3\envs\ADProject\lib\site-packages\torch\utils\data\dataloader.py", line 390, in _get_iterator
    return _MultiProcessingDataLoaderIter(self)
  File "C:\ProgramData\Anaconda3\envs\ADProject\lib\site-packages\torch\utils\data\dataloader.py", line 1077, in __init__
    w.start()
  File "C:\ProgramData\Anaconda3\envs\ADProject\lib\multiprocessing\process.py", line 121, in start
    self._popen = self._Popen(self)
  File "C:\ProgramData\Anaconda3\envs\ADProject\lib\multiprocessing\context.py", line 224, in _Popen
    return _default_context.get_context().Process._Popen(process_obj)
  File "C:\ProgramData\Anaconda3\envs\ADProject\lib\multiprocessing\context.py", line 327, in _Popen
    return Popen(process_obj)
  File "C:\ProgramData\Anaconda3\envs\ADProject\lib\multiprocessing\popen_spawn_win32.py", line 93, in __init__
    reduction.dump(process_obj, to_child)
  File "C:\ProgramData\Anaconda3\envs\ADProject\lib\multiprocessing\reduction.py", line 60, in dump
    ForkingPickler(file, protocol).dump(obj)
AttributeError: Can't pickle local object 'ADMvTec.__init__.<locals>.<lambda>'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "C:\ProgramData\Anaconda3\envs\ADProject\lib\multiprocessing\spawn.py", line 116, in spawn_main
    exitcode = _main(fd, parent_sentinel)
  File "C:\ProgramData\Anaconda3\envs\ADProject\lib\multiprocessing\spawn.py", line 126, in _main
    self = reduction.pickle.load(from_parent)
EOFError: Ran out of input
```

I am not able to figure out a solution to this problem. Can you please let me know how to correct/prevent this problem?
Thank you very much for your help in advance.

**Comment by liznerski** [comment]
Hey, this seems to be again the error mentioned in #2. The code is tested only on Linux and not guaranteed to work on Windows. We have a  [windows](../../tree/windows) branch where we have fixed some errors to make the code Windows compatible. Have you tried that?

**Comment by layadas** [comment]
Thank you for your suggestion. Yes, I did try the windows branch. However, the error was still there. I then figured out that the problem was actually related to PyTorch and CUDA versions. I did a clean install from the wheel files of PyTorch 1.9 by editing the `requirements.txt` file and then it worked. This is an unexpected error that should not have happened according to the `requirements.txt` file in the project, but for some reason it did. This issue is solved now. Thanks again for your suggestion.
